ttnn/cpp/ttnn/operations/experimental/dropout/device/dropout_program_factory.cpp

```diff
@@ -0,0 +1,347 @@
+         /* dropout_kernel_group_1_id = */ kernels.compute_group_1,
+         /* dropout_kernel_group_2_id = */ kernels.compute_group_2,
+         /* core_group_1              = */ core_group_1,
+         /* core_group_2              = */ core_group_2,
+         /* num_cores                 = */ num_cores,
+         /* num_cores_y               = */ num_cores_y}};
+}
+
+void DropoutProgramFactory::override_runtime_arguments(
+    cached_program_t& cached_program,
+    const operation_attributes_t& operation_attributes,
+    const tensor_args_t& tensor_args,
+    tensor_return_value_t& output) {
+    using namespace tt::tt_metal;
+
+    auto& shared_vars = cached_program.shared_variables;
+    auto& dropout_reader_kernel = shared_vars.dropout_reader_kernel_id;
+    auto& dropout_writer_kernel = shared_vars.dropout_writer_kernel_id;
+    auto& dropout_group_1_kernel = shared_vars.dropout_kernel_group_1_id;
+    auto& dropout_group_2_kernel = shared_vars.dropout_kernel_group_2_id;
+    auto& core_group_1 = shared_vars.core_group_1;
+    auto& core_group_2 = shared_vars.core_group_2;
+    auto& program = cached_program.program;
+
+    uint32_t num_cores = shared_vars.num_cores;
+    uint32_t num_cores_y = shared_vars.num_cores_y;
+
+    const auto& input = tensor_args.input;
+    auto src_buffer = input.buffer();
+    auto dst_buffer = output.buffer();
+
+    // Only seed/address arguments need updating here; tile counts remain the same as in create().
+    auto& reader_runtime_args = GetRuntimeArgs(program, dropout_reader_kernel);
+    auto& writer_runtime_args = GetRuntimeArgs(program, dropout_writer_kernel);
+    auto& group_1_runtime_args = GetRuntimeArgs(program, dropout_group_1_kernel);
+    // we need to initialize it with something, but if group 2 is  empty it will be used in the loop
+    auto& group_2_runtime_args =
+        core_group_2.ranges().empty() ? group_1_runtime_args : GetRuntimeArgs(program, dropout_group_2_kernel);
+
+    for (uint32_t i = 0; i < num_cores; i++) {
+        CoreCoord core = {i / num_cores_y, i % num_cores_y};
+
+        // Update the source address for the reader kernel
+        {
+            auto& runtime_args = reader_runtime_args[core.x][core.y];
+            runtime_args[0] = src_buffer->address();
+        }
+        // Update the destination address for the writer kernel
+        {
+            auto& runtime_args = writer_runtime_args[core.x][core.y];
+            runtime_args[0] = dst_buffer->address();
+        }
+        // Update the seed for the compute kernels
+        if (core_group_1.contains(core)) {
+            auto& runtime_args = group_1_runtime_args[core.x][core.y];
+            runtime_args[0] = operation_attributes.seed;
+        } else if (core_group_2.contains(core)) {
+            auto& runtime_args = group_2_runtime_args[core.x][core.y];
+            runtime_args[0] = operation_attributes.seed;
+        } else {
+            TT_THROW("Core not in specified core ranges.");
+        }
```

Comment:
This was also copied from unary code setup which is currently suboptimal (need to review/go over unary code at some point and clean it up 😞).
Ideally we can remove the `contains` checks since we iterate over cores in order, so we only need to know how many cores are in group 1, and can simplify this to an integer comparison.

Ex
```
    uint32_t g1_num_cores = core_group_1.num_cores();
    for (uint32_t i = 0; i < num_cores; i++) {
        CoreCoord core = {i / num_cores_y, i % num_cores_y};
        if (i < g1_num_cores) {
            auto& runtime_args = group_1_runtime_args[core.x][core.y];
            runtime_args[0] = operation_attributes.seed;
        } else {
            auto& runtime_args = group_2_runtime_args[core.x][core.y];
            runtime_args[0] = operation_attributes.seed;
        }
    }
```

Reply:
I see, I'd use contains here. But agree that we need to deal with these two different groups and make code more clear.

Reply:
In general If it is guaranteed that I can do it. I think better to check now.

Reply:
As long as the call to `split_work_to_cores` has the row_wise=true/false arg set the same way we iterate over the cores for rtas, then this would be guaranteed. We do this in some other ops as well. Ex concat calls `auto [num_cores, all_cores, core_group_1, core_group_2, num_tiles_per_core_group_1, num_tiles_per_core_group_2] = tt_metal::split_work_to_cores(compute_with_storage_grid_size, num_output_pages, rm_orientation);` and then generates the core list to iterate over with `const auto cores = grid_to_cores(num_cores, num_cores_x, num_cores_y, rm_orientation);`. In the case for dropout it's iterating over the cores in a hardcoded column-wise fashion, and is calling `split_work_to_cores` and using the default for row_wise arg (false) so would be okay to use the `if (i < g1_num_cores)` check (should probably remove the dependency on the default arg though).

This is a perf optimization so we can leave it as a todo for later if we just want to get a functional version merged.

Reply:
In general I think using CoreRangeSet designated function is a good idea by default.